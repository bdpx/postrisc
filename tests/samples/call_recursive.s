.text
    alloc	47
    write	"test recursive calls"
    ldi.l	%r46, 0x7FFFFFFFFFFFFFFF		; comment
    ldi.l	%r46, 0x8000000000000000
    addi	%r46, %r46, -1
    write	"%i64(r46)"

    mf.spr	%r20, %rsc

    alloc	54		; extend frame to 54 regs
    ldi		%r48, 1		; 
    ldi		%r53, 3		; 1 arg (33+16)
    ldi		%r52, 2		; 2 arg (34+16)
    ldi		%r51, 1		; 3 arg (35+16)
    write	"rsc: %s(rsc)"
    call.r	%r50, func	; call func subroutine, safe 50 regs
    write	"r51=%i64(r51) rsc=%s(rsc)"
    ldi		%r53, 10
    call.r	%r52, rekurs
    write	"rsc: %s(rsc)"
    write	"rsp: %s(rsp)"
;   write	"%m(dump)"
    jmp	smallend
func:
; at entry point func subroutine has 4 regs in frame
    alloc	8   ; extend frame from 4 to 8 regs
    write	"r0      %x128(r0)"		; print packed caller frame and return address
    write	"r1=%i64(r1) r2=%i64(r2) r3=%i64(r3)" ; print args
    ldi		%r1, 12345
    ret

rekurs:
    alloc	4
    write	"r0=%x128(r0) r1=%i64(r1)"
    write	"rsc: %s(rsc)"
    write	"rsp: %s(rsp)"
    addi	%r3, %r1, -1
    ldi		%r2, 0
    b.eq.d	%r1, %r2, rekret
;	cneq	%r1, %r2, 1, 0
    call.r	%r2, rekurs
rekret:
    write	"rsp: %s(rsp)"
    write	"r0: %x128(r0)"
    ret.f	0
smallend:
    nop		0
    nop		111
    alloc	96
    write	"end_call_recursive"
.end
