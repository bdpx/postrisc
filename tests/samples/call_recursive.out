test recursive calls
9223372036854775807
rsc: 0000000000000036 sof=54 bof=0 soc=0 ndirty=0 bod=0
r0      0000000000028232_fffffffff1230060
r1=1 r2=2 r3=3
r51=12345 rsc=0000000000000036 sof=54 bof=0 soc=0 ndirty=0 bod=0
r0=0000000000024134_fffffffff1230070 r1=10
rsc: 0000000034003404 sof=4 bof=52 soc=0 ndirty=52 bod=0
rsp: fffffffff1730000
r0=0000000000010102_fffffffff12300c2 r1=9
rsc: 0000000036003604 sof=4 bof=54 soc=0 ndirty=54 bod=0
rsp: fffffffff1730000
r0=0000000000010102_fffffffff12300c2 r1=8
rsc: 0000000038003804 sof=4 bof=56 soc=0 ndirty=56 bod=0
rsp: fffffffff1730000
r0=0000000000010102_fffffffff12300c2 r1=7
rsc: 000000003a003a04 sof=4 bof=58 soc=0 ndirty=58 bod=0
rsp: fffffffff1730000
r0=0000000000010102_fffffffff12300c2 r1=6
rsc: 000000003c003c04 sof=4 bof=60 soc=0 ndirty=60 bod=0
rsp: fffffffff1730000
r0=0000000000010102_fffffffff12300c2 r1=5
rsc: 000000003e003e04 sof=4 bof=62 soc=0 ndirty=62 bod=0
rsp: fffffffff1730000
r0=0000000000010102_fffffffff12300c2 r1=4
rsc: 0000000040004004 sof=4 bof=64 soc=0 ndirty=64 bod=0
rsp: fffffffff1730000
r0=0000000000010102_fffffffff12300c2 r1=3
rsc: 0000000042004204 sof=4 bof=66 soc=0 ndirty=66 bod=0
rsp: fffffffff1730000
r0=0000000000010102_fffffffff12300c2 r1=2
rsc: 0000000044004404 sof=4 bof=68 soc=0 ndirty=68 bod=0
rsp: fffffffff1730000
r0=0000000000010102_fffffffff12300c2 r1=1
rsc: 0000000046004604 sof=4 bof=70 soc=0 ndirty=70 bod=0
rsp: fffffffff1730000
r0=0000000000010102_fffffffff12300c2 r1=0
rsc: 0000000048004804 sof=4 bof=72 soc=0 ndirty=72 bod=0
rsp: fffffffff1730000
rsp: fffffffff1730000
r0: 0000000000010102_fffffffff12300c2
rsp: fffffffff1730000
r0: 0000000000010102_fffffffff12300c2
rsp: fffffffff1730000
r0: 0000000000010102_fffffffff12300c2
rsp: fffffffff1730000
r0: 0000000000010102_fffffffff12300c2
rsp: fffffffff1730000
r0: 0000000000010102_fffffffff12300c2
rsp: fffffffff1730000
r0: 0000000000010102_fffffffff12300c2
rsp: fffffffff1730000
r0: 0000000000010102_fffffffff12300c2
rsp: fffffffff1730000
r0: 0000000000010102_fffffffff12300c2
rsp: fffffffff1730000
r0: 0000000000010102_fffffffff12300c2
rsp: fffffffff1730000
r0: 0000000000010102_fffffffff12300c2
rsp: fffffffff1730000
r0: 0000000000024134_fffffffff1230070
rsc: 0000000000000036 sof=54 bof=0 soc=0 ndirty=0 bod=0
rsp: fffffffff1730000
end_call_recursive
